edgedb-protocol/src/descriptors.rs

```diff
@@ -569,34 +692,141 @@ impl Decode for ShapeElement {
             (buf.get_u8() as u32, None)
         };
         let name = String::decode(buf)?;
-        ensure!(buf.remaining() >= 2, errors::Underflow);
-        let type_pos = TypePos(buf.get_u16());
+        let type_pos = TypePos::decode(buf)?;
+        let source_type_pos = if buf.proto().is_at_least(2, 0) {
+            Some(TypePos::decode(buf)?)
+        } else {
+            None
+        };
         Ok(ShapeElement {
             flag_implicit: flags & 0b001 != 0,
             flag_link_property: flags & 0b010 != 0,
             flag_link: flags & 0b100 != 0,
             cardinality,
             name,
             type_pos,
+            source_type_pos,
+        })
+    }
+}
+
+impl Decode for InputShapeTypeDescriptor {
+    fn decode(buf: &mut Input) -> Result<Self, DecodeError> {
+        ensure!(buf.remaining() >= 19, errors::Underflow);
+        assert!(buf.get_u8() == 8);
```

Comment:
As above, this would be clearer if it were `InputShapeTypeDescriptor::DESCRIPTOR_ID` or something similar.